nvflare/private/fed/server/client_manager.py

```diff
@@ -141,9 +141,9 @@ def authenticated_client(self, request, context) -> Optional[Client]:
                 )
                 return None
 
-            for token, client in self.clients.items():
-                if client.name == client_name:
-                    with self.lock:
+            with self.lock:
+                for token, client in self.clients.items():
+                    if client.name == client_name:
                         self.clients.pop(token)
```

Comment:
No, this code is problematic! You cannot add or remove items of self.clients while iterating thru the dict, or you will get an exception. The right way is to first iterate self.clients to build the list of items to be removed; then loop thru this list to pop the items from self.clients dict.

Reply:
fixed.